### PR TITLE
[NGPDCP-2595] PUDO component

### PR DESCRIPTION
**Pull request by kaiceresy**
## Description

**1) Changes**

Add PUDO component for the H5 project.

https://comfortdelgrotaxi.atlassian.net/browse/NGPDCP-2595

image
image
image

**2) Impact**

No

## Check list

- [x] 1. Did you start and verify your changes?
- [x] 2. Did you run build to make sure that your changes can be built successfully?
- [x] 3. No !important flag, inline style in css
- [x] 4. No boilerplate codes (1)
- [x] 5. No duplicate code that exist in common functions?
- [x] 6. All of defined variables should have default value, check null and undefined?
- [x] 7. Use meaningful name for variables, no suffix 1,2,... Describes reference information for easier to understand what's inside. (2)
- [x] 9. Unsubscribed all of subscribers and even listeners when you don't need them.

```
(1)
Boilerplate codes is duplicated multiple times a bunch of the same code. This should be a common function to reuse through your code or you can use generic class / methods or design pattern to avoid the Boilerplate codes.
```

```
(2)
Follow theo coding convention
NO acronym variable name:
WRONG: el, elm
RIGHT: element
Your code can be read as a sentence
```

### CODE REVIEW

**Comment by hoapn-fsoft** [comment]
image
image

some errors related to type

**Comment by hoapn-fsoft** [comment]
image
When I swap 2 destination, always focus on the 2nd item

File: packages/react-compass/src/pudo/pudo-item.tsx [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        )
+
+      default:
+        return <></>
+    }
+  }, [type, value, onValueChange])
+
+  return (
+    <StyledPUDOItem
+      css={{...css, zIndex: itemsLength - 1 - index ?? undefined}}
```
Comment (hoapn-fsoft) [inline packages/react-compass/src/pudo/pudo-item.tsx:68]: using useMemo

File: packages/react-compass/src/pudo/pudo.tsx [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+
+    const newPudoObjValues = newPudoArrValues.reduce(
+      (obj, {name, value}) => ({...obj, [name]: value}),
+      {},
+    ) as Record<TItemKeys, string>
+
+    setPudoItems(newPudoItems)
+    setPudoValues(newPudoObjValues)
+    setArrPudoValues(newPudoArrValues)
+    onValuesChange?.(newPudoObjValues, newPudoArrValues)
```
Comment (hoapn-fsoft) [inline packages/react-compass/src/pudo/pudo.tsx:152]: using debounce

**Review by hoapn-fsoft — CHANGES_REQUESTED** [review]

**Comment by hoapn-fsoft** [comment]
> > image When I swap 2 destination, always focus on the 2nd item
> 
> So what are your expectations? Always focus on the last item or sth more dynamic?
> 
> According to the input UI/order, users tend to write/input from top to bottom items. If they do swap the 2nd with the 3rd item, there are higher chances that the 2nd input is empty, and if users need to update the value of the swap pair, again top to bottom. So keep focusing on the new 2nd item that makes sense and less code tho.

Don't focus on any item when the user swaps. Focus is only shown when the user is entering that item

**Review by pieliedie — APPROVED** [review]

**Review by hoapn-fsoft — APPROVED** [review]

**Pull request merged**